app/jobs/scheduled/bookmark_reminder_notifications.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+module Jobs
+
+  # Runs periodically to send out bookmark reminders, capped at 300 at a time.
+  # Any leftovers will be caught in the next run, because the reminder_at column
+  # is set to NULL once a reminder has been sent.
+  class BookmarkReminderNotifications < ::Jobs::Scheduled
+    MAX_REMINDER_NOTIFICATIONS_PER_RUN = 300
+    JOB_RUN_NUMBER_KEY = 'jobs_bookmark_reminder_notifications_job_run_num'.freeze
+    AT_DESKTOP_CONSISTENCY_RUN_NUMBER = 6
+
+    every 5.minutes
+
+    def execute(args = nil)
+      return if !SiteSetting.enable_bookmarks_with_reminders?
+
+      bookmarks = Bookmark.pending_reminders
+        .where.not(reminder_type: Bookmark.reminder_types[:at_desktop])
+        .includes(:user).order('reminder_at ASC')
+
+      bookmarks.limit(MAX_REMINDER_NOTIFICATIONS_PER_RUN).each do |bookmark|
+        BookmarkReminderNotificationHandler.send_notification(bookmark)
+      end
+
+      # we only want to ensure the desktop consistency every X runs of this job
+      # (every 30 mins in this case) so we don't bother redis too much, and the
+      # at desktop consistency problem should not really happen unless people
+      # are setting the "at desktop" reminder, going out for milk, and never coming
+      # back
+      current_job_run_number = Discourse.redis.get(JOB_RUN_NUMBER_KEY)&.to_i || 0
+      if current_job_run_number == AT_DESKTOP_CONSISTENCY_RUN_NUMBER
+        ensure_at_desktop_consistency
+      end
+
+      increment_job_run_number(current_job_run_number)
+    end
+
+    def increment_job_run_number(current_job_run_number)
+      if current_job_run_number.zero? || current_job_run_number == AT_DESKTOP_CONSISTENCY_RUN_NUMBER
+        new_job_run_number = 1
+      else
+        new_job_run_number = current_job_run_number + 1
+      end
+      Discourse.redis.set(JOB_RUN_NUMBER_KEY, new_job_run_number)
+    end
+
+    def ensure_at_desktop_consistency
+      pending_at_desktop_bookmark_reminders = \
+        Bookmark.includes(:user)
+          .references(:user)
+          .pending_at_desktop_reminders
+          .where('users.last_seen_at >= :one_day_ago', one_day_ago: Time.now.utc - 24.hours)
+
+      return if pending_at_desktop_bookmark_reminders.count.zero?
+
+      unique_users = pending_at_desktop_bookmark_reminders.map(&:user).uniq
+      pending_reminders_for_redis_check = unique_users.map do |user|
+        "#{BookmarkReminderNotificationHandler::PENDING_AT_DESKTOP_KEY_PREFIX}#{user.id}"
+      end
+
+      Discourse.redis.mget(pending_reminders_for_redis_check).each.with_index do |value, idx|
+        next if value.present?
+        user_id = pending_reminders_for_redis_check[idx][/\d+/].to_i
+        user = unique_users.find { |u| u.id == user_id }
```

Comment:
This involves looping through all the `unique_users` for each iteration. If that array is only ever small it's probably fine, but if you expect more than a handful of users to be in it, I recommend converting `unique_users` to be a `Hash` so you can look it up in constant time.

Reply:
Great point, I don't expect there to be many but a hash is still so much faster here. I was amazed at how fast hash lookups are when I first found that trick